packages/core/src/interpreter.ts

```diff
@@ -1011,8 +1011,12 @@ export class Interpreter<
 
     childService
       .onDone((doneEvent) => {
-        this.removeChild(childService.id);
-        this.send(toSCXMLEvent(doneEvent as any, { origin: childService.id }));
+        this.scheduler.schedule(() => {
```

Comment:
Any particular reason why this is scheduled now instead of being guarded in the `removeChild`? I can't really think of an edge-case to break it right now but scheduling can potentially introduce edge cases. The problem is that we don't know how many tasks and what kind of tasks are already in the queue - so I'm a little bit afraid that somehow one could create a bizarre situation in which this would crash once again or remove a child that should no longer be removed. Or that changing the timing of the `doneEvent` could change behavior in somebody's code.

Reply:
Also - the fix has only been applied for machines but observables, callbacks, behaviors etc can suffer from the very same problem. Guarding in the `removeChild` would fix all of those cases at once.

Reply:
It's supposed to be scheduled because we need to consider `.onDone(...)` to be an "event", and events need to be queued.

Reply:
Also, if we don't schedule this, there are other potential race conditions like this one, so we need to schedule regardless. I'll fix it for the others.

Reply:
I agree that `this.send` should be scheduled here... but it already is. It already happens in the `send` method itself:
https://github.com/statelyai/xstate/blob/8d7cbc36432c26fb33e485dbddbe41b784de3fc6/packages/core/src/interpreter.ts#L593

I'm pretty sure that we shouldn't schedule this twice - it makes the code execution harder to understand and a little bit iffy. At the moment I see the scheduler as a way to express the external events queue and double scheduling breaks that mental model. Consider this scheduler state:
```
// initial~
[A, B]
// DONE, C arrives
[A, B, scheduled(DONE), C]
// we process 3 items and we end up with
[C, DONE]
```
As we may notice here - the items were processed a little bit out of order. I understand that **conceptually** it doesn't quite matter if we assume that actor communication is async and that done event can arrive at any time. However, this still complicates the mental model and execution path a little bit. I think that in v5 we actually might want to align the implementation closer to the mailbox concept - this actually have already happened in some parts of the codebase, and even on the v4 branch:
https://github.com/statelyai/xstate/blob/f135dfae27c43afcbf52c0960c11c2c4011fe597/packages/core/src/behaviors.ts#L116

So if we follow that in the future then that refactor would change the behavior a little bit - because in the example above DONE would get actually processed before C. I know this is really an edge case and hopefully, such a change wouldn't impact anyone but if we could avoid it then this would still IMHO be preferable.

I see how we might want to `removeChild` together with processing the done event - but then IMHO we should just schedule it and let `this.send` to schedule the event. This would avoid the double scheduling problem that got outlined above while processing both things right after another, as a group of 2 scheduled tasks.

Alternatively `removeChild` could be called when actually processing the done event - as it's sort of its result - but that would be bigger refactor so I wouldn't really recommend this at the moment.

Reply:
I removed the double-schedule issue here: https://github.com/statelyai/xstate/pull/2508/commits/a0a835bd00a6264814aa4006847666fec72e53f4